### ISSUE TITLE
Error building Gloo in Caffe2

### ISSUE THREAD

**Issue by Luo-Liang** [facebookincubator/gloo#100]
Hi, when I'm following the instruction here https://caffe2.ai/docs/getting-started.html?platform=centos&configuration=cloud,

I ran into the following errors when running make.

CMake Error at gloo_cuda_generated_nccl.cu.o.Release.cmake:203 (message):
  Error generating
  /sampa/home/liangluo/caffe2/build/third_party/gloo/gloo/CMakeFiles/gloo_cuda.dir/nccl/./gloo_cuda_generated_nccl.cu.o

CMake Error at gloo_cuda_generated_cuda.cu.o.Release.cmake:203 (message):
  Error generating
  /sampa/home/liangluo/caffe2/build/third_party/gloo/gloo/CMakeFiles/gloo_cuda.dir//./gloo_cuda_generated_cuda.cu.o

CMake Error at gloo_cuda_generated_cuda_private.cu.o.Release.cmake:203 (message):
  Error generating
  /sampa/home/liangluo/caffe2/build/third_party/gloo/gloo/CMakeFiles/gloo_cuda.dir//./gloo_cuda_generated_cuda_private.cu.o

make[2]: *** [third_party/gloo/gloo/CMakeFiles/gloo_cuda.dir/nccl/gloo_cuda_generated_nccl.cu.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [third_party/gloo/gloo/CMakeFiles/gloo_cuda.dir/gloo_cuda_generated_cuda.cu.o] Error 1
make[2]: *** [third_party/gloo/gloo/CMakeFiles/gloo_cuda.dir/gloo_cuda_generated_cuda_private.cu.o] Error 1
make[1]: *** [third_party/gloo/gloo/CMakeFiles/gloo_cuda.dir/all] Error 2

Any idea what this is about? Thanks a lot!

**Comment by pietern** [comment]
The output you give doesn't include any clues. Just the message "Error generating XYZ". Googling for this only gives examples where nvcc crashes. Do you have additional output, perhaps on stderr?

**Comment by Luo-Liang** [comment]
Thanks. Any idea on where to look for those errors? That's all i have on Std outputs.

**Comment by Luo-Liang** [comment]
[ 13%] Building NVCC (Device) object third_party/gloo/gloo/CMakeFiles/gloo_cuda.dir/nccl/gloo_cuda_generated_nccl.cu.o
nvcc fatal   : A single input file is required for a non-link phase when an outputfile is specified
CMake Error at gloo_cuda_generated_nccl.cu.o.Release.cmake:203 (message):
  Error generating
  /sampa/home/liangluo/caffe2/build/third_party/gloo/gloo/CMakeFiles/gloo_cuda.dir/nccl/./gloo_cuda_generated_nccl.cu.o

**Comment by Luo-Liang** [comment]
Does this one help?

**Comment by pietern** [comment]
Yes it does. Do you try to build through Conda?

**Comment by Luo-Liang** [comment]
I invoke the build process using the script provided by caffe2. Gloo is in the third-party folder of caffe2 root folder.

**Comment by pietern** [comment]
Do you have Conda installed and in your PATH?

**Comment by Luo-Liang** [comment]
no. is that a prerequisite?

**Comment by pietern** [comment]
No, not at all, but I have seen this error before only when NVCC (the CUDA compiler) was used through Conda.

**Comment by Luo-Liang** [comment]
I see. Do you know how I can get it to print more details?

**Comment by Luo-Liang** [comment]
btw im using centos7

**Comment by pietern** [comment]
@Luo-Liang That you're using CentOS 7 can be relevant. We currently don't do continuous builds for that OS so can't guarantee it works with the latest revision.

Could you retry this setup and reconfigure CMake with `-DCMAKE_VERBOSE_MAKEFILE=1`? Then CMake will echo the full compiler command before executing it. Hopefully it will give us some more details.

**Comment by giantpoplar** [comment]
hi, @Luo-Liang ,have you solved this problem?

**Comment by giantpoplar** [comment]
hi, @Luo-Liang ,  @pietern  ,is this problem solved now?

**Comment by harsv** [comment]
Hi,
I am having similar issue, some of the include paths have missing "-I" at the beginning. How should I solve this issue. I have centos7+cuda8.0+python2.7 (no conda)

**Comment by pietern** [comment]
I don't know how to solve this yet -- we're not running CentOS builds in our testing environment (both Caffe2 and Gloo), only Ubuntu. We'd like to add CentOS to the mix and when we do will likely run into the same issue.. I'll work on this for a bit, but can't promise I'll figure it out quickly :)

**Comment by akyrola** [comment]
I have the same issue... This is the error I get when running make install 
https://gist.github.com/akyrola/642c89be9bdcf057c3c5436389cce875

**Comment by akyrola** [comment]
I have fresh CentOS 7 install, CUDA 9.

**Comment by akyrola** [comment]
Looks like the problem is with Cmake 3.6.4 (which yum installs). Using CMake 3.10 version fixes the issue. I found this from another thread: https://github.com/caffe2/caffe2/issues/1397

@pietern FYI

**Comment by pietern** [comment]
@akyrola Thanks! I wasn't able to repro on CentOS 7 but that was with older CMake so that makes sense. Where did you get CMake 3.6.4 installed?

**Comment by pietern** [comment]
FWIW, the default cmake for me is 2.8.12. I can only run CMake 3.6.3 by installing "cmake3":

```
$ cmake --version | head -1
cmake version 2.8.12.2
$ cmake3 --version | head -1
cmake3 version 3.6.3
$ cat /etc/os-release 
NAME="CentOS Linux"
VERSION="7 (Core)"
```

**Comment by pietern** [comment]
Just tried with cmake 3.6.3 but doesn't fail at all.

What CentOS image did you use to get a copy of 3.6.4?

**Comment by harsv** [comment]
I still get those errors with cmake3.6.3:

$cat /etc/centos-release
CentOS Linux release 7.4.1708 (Core) 
$ cmake3 --version | head -1
cmake3 version 3.6.3

However, It is gone with cmake3.10

**Comment by pietern** [comment]
@harsv What is your CUDA/nvcc version? 

I'm on 7.4.1708 as well and cannot repro with CMake 3.6.3.

```
# nvcc --version 
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2017 NVIDIA Corporation
Built on Fri_Sep__1_21:08:03_CDT_2017
Cuda compilation tools, release 9.0, V9.0.176
```

**Comment by harsv** [comment]
I am using 8.0, I can also try 9.0 and let you know:
nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2016 NVIDIA Corporation
Built on Tue_Jan_10_13:22:03_CST_2017
Cuda compilation tools, release 8.0, V8.0.61

**Comment by pietern** [comment]
I tried 8.0 as well and can't repro the problem with that either. I ran `rpm -qa | sort` to get a list of installed packages so we can see what the difference is. It is available [here](https://gist.github.com/pietern/dddbb73ad5743e5bbc92f0029b7bd219). Can you do the same and post your package list?

It's nice to know upgrading CMake solves the problem, but I rather figure out the underlying cause before it shapeshifts and comes back in different form...

**Comment by harsv** [comment]
I put the rpm-qa output in comments.
Looks like, I have a bit different setup. Also I have cuda,opemnpi,opencv available as modules.

**Comment by pietern** [comment]
@harsv Thanks for posting. Now we can hopefully find some difference and repro.

The list of packages I'm using comes from the Docker image at `ci.pytorch.org/caffe2/py2-cuda8.0-cudnn7-centos7:25` (you can `docker run` this to get into the environment). This is why the list is much smaller and doesn't include desktop packages. Hopefully the matching packages show some version differences that lead us in the right direction.

**Comment by pietern** [comment]
One difference stands out:
```
A: epel-release-7-9.noarch
B: epel-release-7-11.noarch
```

I'll try and upgrade our CentOS images to use a newer EPEL. I'm confident that will lead to a repro.
